File: System.cpp
#include "System.hpp"
#include "lualib/lua.hpp"

System::System() : devicesAttached(0) {}

bool System::init() {
  return true;
}

Device *System::getDevice(int port) {
  return this->devices[port];
}

bool System::addDevice(int port, Device *device, Setting &sett) {
  if (device->init(sett) == 1) {
    return false;
  }
  this->devices[port] = device;
  this->devicesAttached += 1;
  return true;
}

bool System::removeDevice(int port, bool shouldDelete) {
  if (this->devices[port]->deinit() == 1) {
    return false;
  }
  if (shouldDelete) {
    delete this->devices[port];
  }
  this->devices[port] = NULL;
  this->devicesAttached -= 1;
  return true;
}

File: hardware/system.lua
local system = {}

local devices = {}

function system.getDevice(port)
  return devices[port]
end

function system.getDevices()
  local devs
  for i,v in devices do
    devs[i] = v
  end
  return devs
end

function system.addDevice(device)
  table.insert(devices, device)
end

function system.removeDevice(port)
  if devices[port] == nil then
    return false
  end
  devices[port] = nil
  return true
end

return system
